atlasdb-autobatch/src/test/java/com/palantir/atlasdb/autobatch/CoalescingBatchingEventHandlerTests.java

```diff
@@ -0,0 +1,212 @@
+                .isEqualTo(defaultResponse);
+    }
+
+    @Test
+    public void batchesAreRespected() throws ExecutionException, InterruptedException {
+        Request request = ImmutableRequest.of(5);
+        Response firstInvocation = ImmutableResponse.of(-5);
+        Response secondInvocation = ImmutableResponse.of(-10);
+
+        when(function.apply(ImmutableSet.of(request)))
+                .thenReturn(ImmutableMap.of(request, firstInvocation))
+                .thenReturn(ImmutableMap.of(request, secondInvocation))
+                .thenThrow(new AssertionError("should not reach here"));
+
+        CoalescingBatchingEventHandler<Request, Response> handler = new CoalescingBatchingEventHandler<>(function, 5);
+        Future<Response> firstInvocationResponse = addAndEndBatch(handler, request);
+        Future<Response> secondInvocationResponse = addAndEndBatch(handler, request);
+
+        assertThat(firstInvocationResponse.get())
+                .as("first invocation of the same request gets first response")
+                .isEqualTo(firstInvocation);
+
+        assertThat(secondInvocationResponse.get())
+                .as("second invocation of the same request gets second response")
+                .isEqualTo(secondInvocation);
+    }
+
+    @Test
+    public void exceptionsFailUncompletedElementsInBatch() {
+        RuntimeException exception = new RuntimeException("something went wrong");
+
+        when(function.apply(anySet()))
+                .thenThrow(exception);
+
+        CoalescingBatchingEventHandler<Request, Response> handler = new CoalescingBatchingEventHandler<>(function, 5);
+        Future<Response> firstInvocationResponse = addToBatch(handler, ImmutableRequest.of(5));
+        Future<Response> secondInvocationResponse = addAndEndBatch(handler, ImmutableRequest.of(10));
+
+        assertThatExceptionOfType(ExecutionException.class)
+                .isThrownBy(firstInvocationResponse::get)
+                .withCause(exception);
+
+        assertThatExceptionOfType(ExecutionException.class)
+                .isThrownBy(secondInvocationResponse::get)
+                .withCause(exception);
+    }
+
+    private static Future<Response> addToBatch(
+            CoalescingBatchingEventHandler<Request, Response> handler,
+            Request request) {
+        return addElementToBatch(handler, request, false);
+    }
+
+    private static Future<Response> addAndEndBatch(
+            CoalescingBatchingEventHandler<Request, Response> handler,
+            Request request) {
+        return addElementToBatch(handler, request, true);
+    }
+
+    private static Future<Response> addElementToBatch(
+            CoalescingBatchingEventHandler<Request, Response> handler,
+            Request request,
+            boolean endBatch) {
+        TestBatchElement element = ImmutableTestBatchElement.of(request);
+        handler.onEvent(element, new Random().nextLong(), endBatch);
```

Comment:
Is `new Random().nextLong()` really ok for sequence numbers? I would have expected them to be increasing - but guess this works for testing

Reply:
unused, this would be filled in by the disruptor framework, can add an incrementing counter if you'd wish